proposals/0000-no-toplevel-field-selectors.rst

```diff
@@ -0,0 +1,134 @@
+
+Record definition no longer defines toplevel selector functions when the
+``NoToplevelFieldSelectors`` extension is enabled. A
+``NoToplevelFieldSelectors`` pragma for records will also be available, such
+that this extension can be enabled on a per-record basis.
+
+Record construction/update syntax and pattern matching will work as before, the
+disambiguation handled by ``DuplicateRecordFields``. The necessary selectors
+will be part of a ``Record(..)`` export, or can also be named one by one. They
+always have to be associated with the data type though, because there is no more
+toplevel selector.
+
+A function for Template Haskell to go from a record field name to a selector for
+that field will be provided, because it's not possible anymore to go from field
+name directly to selector.
+
+Example
+^^^^^^^
+
+Given a data structure
+
+    data Foo = Foo { bar :: Int, baz :: String }
+
+The following will be available:
+
+- the type ``Foo``
+- the constructor ``Foo``
+- the two functions ``bar`` and ``baz``
+- the names ``bar`` and ``baz`` for record construction (``Foo { bar = 3, baz = "foo" }``)
+- the names ``bar`` and ``baz`` for ``RecordWildCards``
+
+If the language extension ``NoToplevelFieldSelectors`` is enabled for the module
+or ``Foo`` specifically, all of the above will be generated, except for the two
+functions ``bar`` and ``baz``.
+
+Wildcard exports will work as before, except for the two functions. Even if
+these functions are otherwise defined, the wildcard will not export them.
+Exporting the names for record construction now has to be specific to the
+record. Without ambiguitiy, previously this was equivalent
+
+    module A where (Foo(Foo, bar, baz))
+
+    module A where (Foo(Foo, bar), baz)
+
+Because of the new semantics, these two export statements are now different.
+Maybe a wildcard export will export functions of the same name as well, this
+question isn't resolved yet.
+
+Effect and Interactions
+-----------------------
+
+Breakage estimation
+^^^^^^^^^^^^^^^^^^^
+
+Enabling this extension will break a lot of Template Haskell. Going from record
+field name to selector won't work anymore. A new way to go from record field
+name to selector has to be found.
+
+Anything that generates code with the help of Generic should be fine. The same
+functionality that generates the anonymous functions for Generic could be used
+to provide TH functionality to replace the existing toplevel functions.
+
+The record extensions NamedFieldPuns, RecordWildCards, DisambiguateRecordFields,
+and DuplicateRecordFields are unaffected by this change.
+
+
+Costs and Drawbacks
+-------------------
+
+This might cause some confusion that record fields can't be accessed by toplevel
+selectors anymore - however, that shouldn't be too big of an issue, because some
+library authors already stopped exporting these selectors so they don't have to
+break downstream software on record changes.
+
+
+Alternatives
+------------
+
+None.
```

Comment:
As specified, this extension applies at the definition sites of fields. One could imagine making it apply at use sites instead, so that a module could take selector names out of scope (as variables in expressions), even if they come from imported modules that don't use the extension. Is that worth considering?

Reply:
Then you'd need an orphan `Generic` instance, or to enable one of the language extensions to use them. Might be an interesting addition to use old datatypes with this approach. When creating this library, I was mostly thinking of where you have control over the declaration sites, not where you don't. I'd still go with this proposal, and consider this to be an additional feature.